### PR TITLE
[CARBONDATA-1524][CARBONDATA-1525][PreAgg] Added support for aggregate table drop

### PR DESCRIPTION
**Pull request by kunal642**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[CARBONDATA-<Jira issue #>] Description of pull request`
   
 - [X] Make sure to add PR description including
        
        - the root cause/problem statement
        - What is the implemented solution

 - [X] Any interfaces changed?
 Metastore interface is changed to support child table drop

 - [X] Any backward compatibility impacted?
 No
 - [X] Document update required?
No
 - [X] Testing done
 Test cases are added for child and parent table drop scenarios.
         
 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
                 
---

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/706/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1342/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/709/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1345/

**Comment by jackylk** [comment]
Please tag the PR title with [AggTable]

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/725/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1362/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/848/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/849/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/850/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/861/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1481/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/883/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/888/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1498/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/892/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1503/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1507/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/914/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/915/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/916/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/919/

**Comment by kunal642** [comment]
@ravipesala Please review

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/CarbonSessionExample.scala [hunk 1]
```diff
@@ -56,7 +56,8 @@ object CarbonSessionExample {
       s"""
          | LOAD DATA LOCAL INPATH '$path'
          | INTO TABLE carbon_table
-         | OPTIONS('HEADER'='true', 'COMPLEX_DELIMITER_LEVEL_1'='#')
+         | OPTIONS('FILEHEADER'='shortField,intField,bigintField,doubleField,stringField,timestampField,decimalField,dateField,charField,floatField,complexData',
```
Comment (ravipesala) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/CarbonSessionExample.scala]: don't change

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateDrop.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+  test("dropping 1 aggregate table should not drop others") {
+    sql(
+      "create table preagg1 stored BY 'carbondata' tblproperties('parent'='maintable') as select" +
+      " a,sum(b) from maintable group by a")
+    sql(
+      "create table preagg2 stored BY 'carbondata' tblproperties('parent'='maintable') as select" +
+      " a,sum(c) from maintable group by a")
+    sql("drop table if exists preagg2")
+    checkExistence(sql("show tables"), false, "preagg2")
```
Comment (ravipesala) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateDrop.scala]: Add check for other preagg1 exists

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1529/

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala [hunk 3]
```diff
@@ -446,6 +447,35 @@ class CarbonFileMetastore extends CarbonMetaStore {
...
 
+  protected def updateParentTableInfo(parentRelationIdentifier: RelationIdentifier,
+      childCarbonTable: CarbonTable)(sparkSession: SparkSession): Unit = {
+    val dbName = parentRelationIdentifier.getDatabaseName
+    val tableName = parentRelationIdentifier.getTableName
+    try {
+      val parentCarbonTable = CarbonEnv.getInstance(sparkSession).carbonMetastore
+        .lookupRelation(Some(dbName), tableName)(sparkSession).asInstanceOf[CarbonRelation]
+        .tableMeta.carbonTable
+      val childSchemas = parentCarbonTable.getTableInfo.getDataMapSchemaList
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala]: add a null check and throw exception saying unexpected

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala [hunk 4]
```diff
@@ -446,6 +447,35 @@ class CarbonFileMetastore extends CarbonMetaStore {
...
+      val schemaConverter = new ThriftWrapperSchemaConverterImpl
+      PreAggregateUtil
+        .updateSchemaInfo(parentCarbonTable,
+          schemaConverter
+            .fromWrapperToExternalTableInfo(parentCarbonTable.getTableInfo,
+              dbName,
+              tableName))(sparkSession)
+    } catch {
+      case ex: Exception =>
+        LOGGER.error(ex, s"Table $dbName.$tableName does not exist.")
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala]: throw exception

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1530/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1531/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/932/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1534/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1547/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/949/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1564/

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/Compactor.scala [hunk 5]
```diff
@@ -89,23 +89,23 @@ object Compactor {
...
-    } else {
-      new CarbonMergerRDD(
-        sc.sparkContext,
-        new MergeResultImpl(),
-        carbonLoadModel,
-        carbonMergerMapping,
-        execInstance
-      ).collect
-    }
+      if (compactionType == CompactionType.IUD_UPDDEL_DELTA_COMPACTION) {
```
Comment (ravipesala) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/Compactor.scala]: don't format if you don't change the class

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 6]
```diff
@@ -33,8 +33,10 @@ import org.apache.carbondata.core.metadata.CarbonTableIdentifier
 import org.apache.carbondata.core.metadata.datatype.{DataType, DataTypes, DecimalType}
 import org.apache.carbondata.core.metadata.encoder.Encoding
 import org.apache.carbondata.core.metadata.schema._
-import org.apache.carbondata.core.metadata.schema.table.{CarbonTable, RelationIdentifier, TableInfo, TableSchema}
-import org.apache.carbondata.core.metadata.schema.table.column.{ColumnSchema, ParentColumnTableRelation}
+import org.apache.carbondata.core.metadata.schema.table.{CarbonTable, RelationIdentifier,
```
Comment (ravipesala) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: Don't change the class

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/CarbonDropTableCommand.scala [hunk 7]
```diff
@@ -60,33 +62,40 @@ case class CarbonDropTableCommand(
         lock => carbonLocks += CarbonLockUtil.getLockObject(carbonTableIdentifier, lock)
       }
       LOGGER.audit(s"Deleting table [$tableName] under database [$dbName]")
-
-      // fires the event before dropping main table
-      val carbonTable = CarbonMetadata.getInstance.getCarbonTable(dbName + "_" + tableName)
+      val carbonTable: Option[CarbonTable] = try {
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/CarbonDropTableCommand.scala]: try get from cache using `getTableFromMetadataCache` if not get then do lookup.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/CarbonDropTableCommand.scala [hunk 8]
```diff
@@ -60,33 +62,40 @@ case class CarbonDropTableCommand(
...
+          if (!ifExistsSet) {
+            throw ex
+          }
+          None
+      }
       val operationContext = new OperationContext
       val dropTablePreEvent: DropTablePreEvent =
         DropTablePreEvent(
-          carbonTable,
+          carbonTable.get,
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/CarbonDropTableCommand.scala]: you should not call get directly as in above catch there is chance of `None`

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/955/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1568/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/957/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1569/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1570/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/958/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1572/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/CarbonDropTableCommand.scala (hunk 8 above)
Comment (kunal642) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/CarbonDropTableCommand.scala]: @fixed

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1573/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/965/

**Comment by ravipesala** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/967/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1579/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1004/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1621/

**Pull request closed without merging**